docs: #1562 doc support for DeleteRecords (#1657)
redpanda-data/documentation: versioned_docs/version-23.2/develop/config-topics.mdx

rpk topic  delete -r '^f.*' '.*r$'
```

Note that the first regular expression starts with the `^` symbol, and the last ends with the `$` symbol. This requirement helps prevent accidental deletions.

## Delete records from a topic

> Redpanda lets you delete data from the beginning of a partition up to a specific offset (event). The offset represents the true creation time of the event, not the time when it was stored by Redpanda. Deleting records frees up disk space, especially if your producers are pushing more data than you anticipated in your retention plan. Do this when you know that all consumers have read up to that given offset, and the data is no longer needed.

Reviewer: ```suggestion
Redpanda lets you delete data from the beginning of a partition up to a specific offset (event). The offset represents the true creation time of the event, not the time when it was stored by Redpanda. Deleting records frees up disk space, which is especially helpful if your producers are pushing more data than you anticipated in your retention plan. Do this when you know that all consumers have read up to that given offset, and the data is no longer needed.
```

> Delete records from a topic with the `rpk topic trim` command. Use this when you know that all consumers have read up to that given offset, and the data is no longer needed.

Reviewer 2: `rpk topic trim` is new for 23.2? Todo: link to its rpk command reference page

:::note notes

> - To delete records, `cleanup.policy` must be set to `delete` or `delete+compact`.

Reviewer 2: ```suggestion
- To delete records, `cleanup.policy` must be set to `delete` or `compact,delete`.
```

> There are different ways to delete records from a topic, including using the `rpk topic trim` command or using the DeleteRecords Kafka API with Kafka clients. 

Reviewer 3: @micheleRP current plan is to use an alias so you can use either 'rpk topic trim-prefix, or rpk topic trim for short'

cc @r-vasquez

> - Object storage is deleted asynchronously. When DeleteRecords returns, the partition's start offset will have advanced, but the work required to garbage collect delete segments may not be complete.

Reviewer 4: nit: garbage collect "deleted" segments

> - Object storage is deleted asynchronously. When DeleteRecords returns, the partition's start offset will have advanced, but the work required to garbage collect deleted segments may not be complete.

Reviewer 2: ```suggestion
- Object storage is deleted asynchronously. When DeleteRecords returns, the partition's start offset will have advanced, but garbage collection of deleted segments may not be complete.
```

> - After calling DeleteRecords, local storage and object storage may contain data from before the offset. (Redpanda does not truncate segments in half. It deletes as many segments as possible for the last segment, then bumps the start offset.) This data is not visible to clients but could be read by someone with access to the local disk of a Redpanda node. 

Reviewer 4: Remove `for the last segment`

> - After calling DeleteRecords, local storage and object storage may contain data from before the offset. (Redpanda does not truncate segments in half. It deletes as many segments as possible for the last segment, then bumps the start offset.) This data is not visible to clients but could be read by someone with access to the local disk of a Redpanda node. 

Reviewer 4: Change `This data` to `Data before the new start offset`

> - After calling DeleteRecords, local storage and object storage may still contain data for deleted offsets. (Redpanda does not truncate segments, instead it deletes as many whole segments as possible, then bumps the start offset.) Data before the new start offset is not visible to clients but could be read by someone with access to the local disk of a Redpanda node. 

Reviewer 4: ```
 (Redpanda does not truncate segments, instead it deletes as many whole segments as possible, then bumps the start offset.)
 ```
 This is backwards, redpanda first bumps the start offset then attempts to delete as many whole segments in the background.

> - After calling DeleteRecords, local storage and object storage may contain data from before the offset. (Redpanda does not truncate segments in half. It deletes as many segments as possible, then bumps the start offset.) Data before the new start offset is not visible to clients but could be read by someone with access to the local disk of a Redpanda node. 

Reviewer 2: ```suggestion
- After calling DeleteRecords, local storage and object storage may still contain data for deleted offsets. (Redpanda does not truncate segments, instead it deletes as many whole segments as possible, then bumps the start offset.) Data before the new start offset is not visible to clients but could be read by someone with access to the local disk of a Redpanda node. 
```

> - After deleting records, local storage and object storage may still contain data for deleted offsets. (Redpanda does not truncate segments, instead it bumps the start offset then attempts to delete as many whole segments as possible.) Data before the new start offset is not visible to clients but could be read by someone with access to the local disk of a Redpanda node. 

Reviewer 3: @micheleRP I think we should add 'similar to Kafka' or something to explain this is the same behavior in Kafka.

rpk topic  delete -r '^f.*' '.*r$'
```

Note that the first regular expression starts with the `^` symbol, and the last ends with the `$` symbol. This requirement helps prevent accidental deletions.

## Delete records from a topic

> To free up disk space, you can use `kafka-delete-records.sh` to delete data from the beginning of a partition up to a requested offset. This DeleteRecords API lets you delete data with precise control up to a point in time or to a specific business event (offset). This can be important for compliance reasons, because the offset represents the true creation time of the event, not when it was stored by Redpanda. The DeleteRecords API is also useful if time-based message retention is storing more data than anticipated. Use this when you know that all consumers have read up to that given point in time and the data is no longer needed.

Reviewer 4: Theres many ways to do this, i like `kcl` (@twmb 's command line tool), maybe we should give a `kcl` example instead ?

Reviewer 4: I wouldn't say up until a point in time, just mention up until a specific business event (offset). The API doesn't include a time point

Reviewer 4: Didn't we say we wanted to let people know this isn't really a great compliance tool? The TL;DR is that after the request is processed some data you requested to be deleted may still be on disk for an indeterminate amount of time, even though no kafka clients would be able to access them.

Reviewer 4: "The DeleteRecords API is also useful if time-based message retention is storing more data than anticipated" What does this mean?

Reviewer 5: We plan to add this to rpk, with something like `rpk topic trim`

Reviewer 5: We will be releasing the rpk change in the same release that DeleteRecords is added -- I think that's 23.2, so no reason to have a kcl example here. It's a pretty quick turnaround for rpk.

Author: Thanks @graphcareful and @twmb. Please review the latest draft with `rpk topic trim` and approve if it's good, or let me know if this doc PR should hold off for final rpk implementation! cc @mattschumpert

> To delete records, run `kafka-delete-records.sh`:

Reviewer 4: Here's an example with KCL:
```
kcl admin delete-records foo:p0,o120 foo:p1,o3888
```

_NEED EXAMPLE_

:::note notes

> - To delete records, `cleanup.policy` cannot be set to `compact`.

Reviewer 4: it can be set to `compact` you need to set `delete`, you can also have `delete+compact`

> - Local storage is deleted synchronously, but if a broker is offline during the request, it may retain data until it comes back online. 

Reviewer 4: Technically its deleted asynchronously. A background thread that runs on a hard coded timer processes these requests. I wouldn't mention data from an offline node will not be deleted that is obvious

> - Object storage is deleted asynchronously. When DeleteRecords returns, the partition's start offset will have advanced, but the work required to garbage collect delete segments may not be complete.

Reviewer 4: +1

> - After calling DeleteRecords, the Tiered Storage cache may contain data from before the offset. This data is not visible to clients but could be read by someone with access to the local disk of a Redpanda node.

Reviewer 4: Also the same is true for local storage. The reason is because we do not truncate segments in half. So we just delete as many segments as possible then for that last segment, just bump the start offset. Cloud storage works the same way too
